### PR TITLE
PLFM-6275

### PR DESCRIPTION
**Pull request by nickgros**
- Add 'additionalFilters' to TransformSqlWithFacetsRequest object
- Support transforming QueryFilters to Synapse SQL
- Refactor to reuse less complex code

Draft while pending API design review

### CODE REVIEW

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java [hunk 1]
```diff
@@ -15,39 +22,157 @@ public class FacetUtils {
...
-					builder.append(searchCondition);
-				}
-			}
-		}
+
+		String joinedConditions = conditionStrings.stream()
+				// Remove nulls
+				.filter(s -> s != null && !s.isEmpty())
+				// Join with " AND "
+				.collect(Collectors.joining(" AND "));
```
Comment (nickgros) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java:77]: IMO this is much more clear than the imperative style loop

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java [hunk 2]
```diff
@@ -15,39 +22,157 @@ public class FacetUtils {
...
+	 */
+	private static String concatConditionStrings(List<String> conditionStrings) {
+		ValidateArgument.required(conditionStrings, "conditionStrings");
+
 		StringBuilder builder = new StringBuilder();
 		builder.append("(");
 		int initialSize = builder.length(); //length with the "( " included
-		
-		for(FacetRequestColumnModel facetColumn : validatedFacets){
-			if(columNameToIgnore == null || !facetColumn.getColumnName().equals(columNameToIgnore)){ //make sure not to include the ignored column
```
Comment (nickgros) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java:30]: `columnNameToIgnore` is now filtered in `concatFacetSearchConditionStrings`

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java [hunk 3]
```diff
@@ -15,39 +22,157 @@ public class FacetUtils {
 	 * @param columNameToIgnore the name of the column to exclude from the concatenation
 	 * @return the concatenated string or null if there was nothing to concatenate
 	 */
-	public static String concatFacetSearchConditionStrings(List<FacetRequestColumnModel> validatedFacets, String columNameToIgnore){
+	public static String concatFacetSearchConditionStrings(List<FacetRequestColumnModel> validatedFacets, Optional<String> columNameToIgnore){
```
Comment (john-hill) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java:25]: Cannot have Optional as method param
Can overload the signature.
In method can use
Name valueOption = Optional.ofNullable(param)

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java [hunk 4]
```diff
@@ -15,39 +22,157 @@ public class FacetUtils {
...
+
+		return concatConditionStrings(filters.stream().map(FacetUtils::getSearchConditionString).collect(Collectors.toList()));
+	}
+
+	/**
+	 * Concatenates a list of SQL conditional statements with " AND "
+	 * @param conditionStrings
+	 * @return
+	 */
+	private static String concatConditionStrings(List<String> conditionStrings) {
```
Comment (john-hill) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java:66]: This means there is no unit test for this.

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java [hunk 5]
```diff
@@ -15,39 +22,157 @@ public class FacetUtils {
...
+				.filter(facet -> {
+					if (columNameToIgnore.isPresent()) {
+						return !facet.getColumnName().equals(columNameToIgnore.get());
+					} else {
+						return true;
+					}
+				})
+				.collect(Collectors.toList());
+
+		List<String> conditionStrings = facetColumnsWithoutIgnoredColumn.stream().map(FacetRequestColumnModel::getSearchConditionString).collect(Collectors.toList());
```
Comment (john-hill) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java:42]: can this be a single line stream with a single line filter.

Comment (john-hill) [reply lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java:42]: ...filter(Not(facet.getColumnName().equal(columNameToIgnore)))

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java [hunk 6]
```diff
@@ -15,39 +22,157 @@ public class FacetUtils {
...
+		return builder.toString();
+	}
+
+	/**
+	 * Converts a QueryFilter to a Synapse SQL condition string
+	 * @param filter
+	 * @return a standalone condition, such as "( \"foo\" LIKE '%bar%' )"
+	 */
+	public static String getSearchConditionString(QueryFilter filter) {
+		if (filter instanceof ColumnSingleValueQueryFilter) {
```
Comment (john-hill) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/util/FacetUtils.java:95]: Find a way to move this logic out of a static util and into a PoJo with methods like toSql().

**Review by john-hill — CHANGES_REQUESTED** [review]
We think this would be easier and cleaner to implement as an extension of the bundle.

**Comment by john-hill** [comment]
We went another route

**Pull request closed without merging**